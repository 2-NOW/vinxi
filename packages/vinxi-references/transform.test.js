import * as acorn from "acorn";
import tsPlugin from "acorn-typescript";
import { generateCode, parseModule } from "magicast";
import { print, parse as recastParse } from "recast";
import { prettyPrint } from "recast";
import { describe, expect, it } from "vitest";

import {
	decorateExports,
	wrapExports,
	wrapExportsPlugin,
} from "./transform-references.js";

function parse(code) {
	return recastParse(code, {
		parser: {
			parse(source) {
				var comments = [];
				var tokens = [];

				const ast = acorn.Parser.extend(tsPlugin()).parse(source, {
					allowHashBang: true,
					allowImportExportEverywhere: true,
					allowReturnOutsideFunction: true,
					locations: true,
					onComment: comments,
					onToken: tokens,
					// additional options
					ecmaVersion: 2024,
					sourceType: "module",
				});

				if (!ast.comments) {
					ast.comments = comments;
				}
				if (!ast.tokens) {
					ast.tokens = tokens;
				}
				return ast;
			},
		},
	});
}
const js = (args) => prettyPrint(parse(args)).code;

function transformSSR(
	code,
	plugin,
	args = {
		id: "test",
		runtime: {
			module: "test",
			function: "runtimeFunction",
		},
		hash: (s) => s,
		options: {
			ssr: true,
		},
	},
) {
	return js(
		plugin({
			code,
			ast: parse(code),
			...args,
		}),
	);
}

describe("wrap exports", () => {
	it("wraps named export of anonymous function", () => {
		expect(
			transformSSR(`export const namedExport = function() {}`, wrapExports),
		).toMatchInlineSnapshot(`
			"import { runtimeFunction } from \\"test\\";
			export const namedExport = runtimeFunction(function() {}, \\"test\\", \\"namedExport\\");"
		`);
	});

	it("wraps named export of anonymous async function", () => {
		expect(
			transformSSR(
				`export const namedExport = async function() {}`,
				wrapExports,
			),
		).toMatchInlineSnapshot(`
			"import { runtimeFunction } from \\"test\\";
			export const namedExport = runtimeFunction(async function() {}, \\"test\\", \\"namedExport\\");"
		`);
	});

	it("wraps named export of arrow function expression", () => {
		expect(
			transformSSR(`export const namedExport = () => {}`, wrapExports),
		).toMatchInlineSnapshot(`
			"import { runtimeFunction } from \\"test\\";
			export const namedExport = runtimeFunction(() => {}, \\"test\\", \\"namedExport\\");"
		`);
	});

	it("wraps named export of  async arrow function expressions", () => {
		expect(
			transformSSR(`export const namedExport = async () => {}`, wrapExports),
		).toMatchInlineSnapshot(`
			"import { runtimeFunction } from \\"test\\";
			export const namedExport = runtimeFunction(async () => {}, \\"test\\", \\"namedExport\\");"
		`);
	});
});
