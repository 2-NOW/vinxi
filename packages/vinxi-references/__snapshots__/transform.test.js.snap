// Vitest Snapshot v1

exports[`wrap exports > wraps named export of  async arrow function expressions 1`] = `
"import { runtimeFunction } from \\"test\\";
export const namedExport = runtimeFunction(async () => {}, \\"test\\", \\"namedExport\\");"
`;

exports[`wrap exports > wraps named export of anonymous async function 1`] = `
"import { runtimeFunction } from \\"test\\";
export const namedExport = runtimeFunction(async function() {}, \\"test\\", \\"namedExport\\");"
`;

exports[`wrap exports > wraps named export of anonymous function 1`] = `
"import { runtimeFunction } from \\"test\\";
export const namedExport = runtimeFunction(function() {}, \\"test\\", \\"namedExport\\");"
`;

exports[`wrap exports > wraps named export of arrow function expression 1`] = `
"import { runtimeFunction } from \\"test\\";
export const namedExport = runtimeFunction(() => {}, \\"test\\", \\"namedExport\\");"
`;
